rust_snuba/rust_arroyo/src/backends/kafka/mod.rs

```diff
@@ -165,107 +165,113 @@ impl<C: AssignmentCallbacks + Send + Sync> ClientContext for CustomContext<C> {
+    //
+    // in comparison, confluent-kafka-python will execute on_assign, and only call rdkafka.assign
+    // if the callback did not already explicitly call assign.
+    //
+    // if we call rdkafka.assign multiple times, we have seen random AutoOffsetReset errors popping
+    // up in poll(), since we (briefly) assigned invalid offsets to the consumer
+    fn rebalance(
+        &self,
+        base_consumer: &BaseConsumer<Self>,
+        err: RDKafkaRespErr,
+        tpl: &mut TopicPartitionList,
+    ) {
+        match err {
+            RDKafkaRespErr::RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS => {
+                let mut partitions: Vec<Partition> = Vec::new();
+                let mut offsets = self.consumer_offsets.lock().unwrap();
+                for partition in tpl.elements().iter() {
+                    let topic = Topic::new(partition.topic());
+                    let index = partition.partition() as u16;
+                    let arroyo_partition = Partition::new(topic, index);
+                    offsets.remove(&arroyo_partition);
+                    partitions.push(arroyo_partition);
+                }
 
-            for partition in committed_offsets.elements() {
-                let raw_offset = partition.offset().to_raw().unwrap();
+                let committer = OffsetCommitter {
+                    consumer: base_consumer,
+                };
 
-                let topic = Topic::new(partition.topic());
+                // before we give up the assignment, strategies need to flush and commit
+                self.callbacks.on_revoke(committer, partitions);
 
-                if raw_offset >= 0 {
-                    offset_map.insert(
-                        Partition::new(topic, partition.partition() as u16),
-                        raw_offset as u64,
-                    );
-                } else {
-                    // Resolve according to the auto offset reset policy
-                    let (low_watermark, high_watermark) = base_consumer
-                        .fetch_watermarks(partition.topic(), partition.partition(), None)
-                        .unwrap();
-
-                    let resolved_offset = match self.initial_offset_reset {
-                        InitialOffset::Earliest => low_watermark,
-                        InitialOffset::Latest => high_watermark,
-                        InitialOffset::Error => {
-                            panic!("received unexpected offset");
+                base_consumer
+                    .unassign()
+                    .expect("failed to revoke partitions");
+            }
+            RDKafkaRespErr::RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS => {
+                let committed_offsets = base_consumer
+                    .committed_offsets((*tpl).clone(), None)
+                    .unwrap();
+
+                let mut offset_map: HashMap<Partition, u64> = HashMap::new();
+                let mut tpl = TopicPartitionList::with_capacity(offset_map.len());
```

Comment:
The `with_capacity` is useless here, as the map is empty. Both can be pre-allocated with the length of `committed_offsets` though.

Reply:
copy-paste error, thanks